From 333e582e651efaf3678ed8b89cd2274f75818649 Mon Sep 17 00:00:00 2001
From: Kevin Modzelewski <kmod@dropbox.com>
Date: Fri, 21 Mar 2014 16:56:28 -0700
Subject: [PATCH] Update TailCallElim to avoid making redundant changes

---
 lib/Transforms/Scalar/TailRecursionElimination.cpp |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/lib/Transforms/Scalar/TailRecursionElimination.cpp b/lib/Transforms/Scalar/TailRecursionElimination.cpp
index 351aee0..7300d84 100644
--- a/lib/Transforms/Scalar/TailRecursionElimination.cpp
+++ b/lib/Transforms/Scalar/TailRecursionElimination.cpp
@@ -252,8 +252,10 @@ bool TailCallElim::runOnFunction(Function &F) {
       for (BasicBlock::iterator I = BB->begin(), E = BB->end(); I != E; ++I) {
         if (CallInst *CI = dyn_cast<CallInst>(I)) {
           if (!ACT.UsesAlloca.count(CI)) {
-            CI->setTailCall();
-            MadeChange = true;
+            if (!CI->isTailCall()) {
+              CI->setTailCall();
+              MadeChange = true;
+            }
           }
         }
       }
-- 
1.7.9.5

