diff -ur cffi-1.2.1_orig/c/cglob.c cffi-1.2.1/c/cglob.c
--- cffi-1.2.1_orig/c/cglob.c	2016-02-22 18:48:06.853772302 +0000
+++ cffi-1.2.1/c/cglob.c	2016-02-25 11:02:26.319738004 +0000
@@ -46,7 +46,15 @@
 static PyObject *make_global_var(PyObject *name, CTypeDescrObject *type,
                                  char *addr, gs_fetch_addr_fn fetch_addr)
 {
-    GlobSupportObject *gs = PyObject_New(GlobSupportObject, &GlobSupport_Type);
+    // Pyston change:
+    static int has_registered = 0;    
+    GlobSupportObject *gs;
+    if (!has_registered) {
+        PyType_Ready(&GlobSupport_Type);
+        has_registered = 1;
+    }
+
+    gs = PyObject_New(GlobSupportObject, &GlobSupport_Type);
     if (gs == NULL)
         return NULL;
 
