From e1e415a1738fd296f83d0b1ce7ad8e7a1682d97a Mon Sep 17 00:00:00 2001
From: Marius Wachtler <undingen@gmail.com>
Date: Mon, 8 Jun 2015 17:35:14 +0200
Subject: [PATCH] Pyston change: register types

---
 _mysql.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/_mysql.c b/_mysql.c
index 5b81c79..718d3f8 100644
--- a/_mysql.c
+++ b/_mysql.c
@@ -3086,6 +3086,11 @@ init_mysql(void)
 	_mysql_ResultObject_Type.tp_free = _PyObject_GC_Del;
 #endif
 #endif
+	// Pyston change:
+	if (PyType_Ready(&_mysql_ConnectionObject_Type) < 0)
+		return;
+	if (PyType_Ready(&_mysql_ResultObject_Type) < 0)
+		return;
 
 	if (!(dict = PyModule_GetDict(module))) goto error;
 	if (PyDict_SetItemString(dict, "version_info",
-- 
2.1.4

