diff -ur PyICU-1.0.1_o/_icu.cpp PyICU-1.0.1/_icu.cpp
--- PyICU-1.0.1_o/_icu.cpp	2010-04-02 00:12:54.000000000 +0200
+++ PyICU-1.0.1/_icu.cpp	2015-05-19 15:26:15.131375981 +0200
@@ -193,6 +193,8 @@
     if (!PyArg_ParseTuple(args, "Os", &object, &doc))
         return NULL;
 
+// Pyston change:
+#if 0
     /* constructors */
     if (PyObject_TypeCheck(object, &PyWrapperDescr_Type))
     {
@@ -206,6 +208,14 @@
         ((PyMethodDescrObject *) object)->d_method->ml_doc = strdup(doc);
         Py_RETURN_NONE;
     }
+#else
+    if (PyObject_TypeCheck(object, &PyWrapperDescr_Type) ||
+        PyObject_TypeCheck(object, &PyMethod_Type) ||
+        PyObject_TypeCheck(object, _method_type)) {
+        // Ignore the doc string for now.
+        Py_RETURN_NONE;
+    }
+#endif
 
     /* class methods */
     if (PyObject_TypeCheck(object, &PyCFunction_Type))
