diff --git a/numpy/core/src/multiarray/compiled_base.c b/numpy/core/src/multiarray/compiled_base.c
index b9db3bb..2d24efc 100644
--- a/numpy/core/src/multiarray/compiled_base.c
+++ b/numpy/core/src/multiarray/compiled_base.c
@@ -1337,13 +1337,13 @@ arr_add_docstring(PyObject *NPY_UNUSED(dummy), PyObject *args)
         _ADDDOC(Type, new->tp_doc, new->tp_name);
     }
     else if (_TESTDOC2(MemberDescr)) {
-        _ADDDOC(MemberDescr, new->d_member->doc, new->d_member->name);
+        /* _ADDDOC(MemberDescr, new->d_member->doc, new->d_member->name); */
     }
     else if (_TESTDOC2(GetSetDescr)) {
-        _ADDDOC(GetSetDescr, new->d_getset->doc, new->d_getset->name);
+        /* _ADDDOC(GetSetDescr, new->d_getset->doc, new->d_getset->name); */
     }
     else if (_TESTDOC2(MethodDescr)) {
-        _ADDDOC(MethodDescr, new->d_method->ml_doc, new->d_method->ml_name);
+        /* _ADDDOC(MethodDescr, new->d_method->ml_doc, new->d_method->ml_name); */
     }
     else {
         PyObject *doc_attr;
